sdks/python/apache_beam/io/datastore/v1/query_splitter.py

```diff
@@ -0,0 +1,256 @@
+  # extra region following the last scatter point. Thus, we do not need the
+  # scatter entity for the last region.
+  scatter_query.limit.value = (num_splits - 1) * KEYS_PER_SPLIT
+  googledatastore.helper.add_projection(scatter_query, KEY_PROPERTY_NAME)
+
+  return scatter_query
+
+
+def _get_scatter_keys(datastore, query, num_splits, partition):
+  """Gets a list of split keys given a desired number of splits.
+
+  This list will contain multiple split keys for each split. Only a single split
+  key will be chosen as the split point, however providing multiple keys allows
+  for more uniform sharding.
+
+  Args:
+    numSplits: the number of desired splits.
+    query: the user query.
+    partition: the partition to run the query in.
+    datastore: the client to datastore containing the data.
+  """
+  scatter_point_query = _create_scatter_query(query, num_splits)
+
+  key_splits = []
+  while True:
+    req = datastore_pb2.RunQueryRequest()
+    if partition:
+      req.partition_id.CopyFrom(partition)
+
+    req.query.CopyFrom(scatter_point_query)
+
+    resp = datastore.run_query(req)
+    for entity_result in resp.batch.entity_results:
+      key_splits.append(entity_result.entity.key)
+
+    if resp.batch.more_results != query_pb2.QueryResultBatch.NOT_FINISHED:
+      break
+
+    scatter_point_query.start_cursor = resp.batch.end_cursor
+    scatter_point_query.limit.value -= len(resp.batch.entity_results)
+
+  key_splits.sort(helper.key_comparator)
+  return key_splits
+
+
+def _get_split_key(keys, num_splits):
+  """Given a list of keys and a number of splits find the keys to split on.
+
+  Args:
+    keys: the list of keys.
+    num_splits: the number of splits.
+
+  """
+
+  # If the number of keys is less than the number of splits, we are limited
+  # in the number of splits we can make.
+  if not keys or (len(keys) < (num_splits - 1)):
+    return keys
+
+  # Calculate the number of keys per split. This should be KEYS_PER_SPLIT,
+  # but may be less if there are not KEYS_PER_SPLIT * (numSplits - 1) scatter
+  # entities.
+  #
+  # Consider the following dataset, where - represents an entity and
+  # * represents an entity that is returned as a scatter entity:
+  # ||---*-----*----*-----*-----*------*----*----||
+  # If we want 4 splits in this data, the optimal split would look like:
+  # ||---*-----*----*-----*-----*------*----*----||
+  #            |          |            |
+  # The scatter keys in the last region are not useful to us, so we never
+  # request them:
+  # ||---*-----*----*-----*-----*------*---------||
+  #            |          |            |
+  # With 6 scatter keys we want to set scatter points at indexes: 1, 3, 5.
+  #
+  # We keep this as a float so that any "fractional" keys per split get
+  # distributed throughout the splits and don't make the last split
+  # significantly larger than the rest.
+
+  num_keys_per_split = max(1.0, float(len(keys)) / (num_splits - 1))
```

Comment:
Yes this should be fine, the datastore will only return how much ever it can. Worst case we will have 1 num_keys_per_split, so actual num_splits could be less than what user had asked for.